### DISCUSSION TITLE
Implementing backend db like ES (JSON friendly graph db)

### DISCUSSION THREAD

**Discussion by daniellefisla** [tuProlog/2p-kt#99] [Ideas]
Hi, would it be possible to store all facts, predicates, in a backend solution like ElasticSearch? I noticed the JSON serialization is already implemented, would it be possible to create like an abstraction layer for a backend engine? 

I am trying to figure out the effort and how much would need to be refactored. 
Is there a common data structure and access point class I should look at?

thank you

**Comment by gciatto** [comment]
Hi @daniellefisla,

I'm not an expert about Elastic Search, but I'll assume it's a sort of NO-SQL database management system (DBMS) for storing JSON documents.
Of course, feel free to correct me if I'm wrong.

Depending on what your needs are, solutions may vary from trivial to moderately articulated.
However, assuming some Java/Kotlin client library exists for Elastic Search (and I guess this is the case), one may "simply" create a novel implementation for the `Theory` or `MutableTheory` interfaces (from the `:theory` module), interoperating with Elastic Search behind the scenes to store, retrieve, and remove logic clauses.
That would of course rely on the already-implemented JSON-serialization facilities (from the `:serialize-core` and `:serialize-theory` modules).
So, I guess it is "just" a matter of _(i)_ doing the right queries to implement all methods of the aforementioned interfaces, and _(ii)_ serializing and deserializing clauses accordingly.

Targetting the `Theory` and `MutableTheory` interfaces is strategical because they are what logic (and Prolog) solvers rely upon to solve logic queries (aka support logic programming). Hence, in principle, my guess is that it would be possible to let a logic solver (e.g. Prolog) transparently compute on top of the DBMS backend of choice. 

A prior bachelor's student of mine did a similar experiment with Neo4J, a few years ago.
Unfortunately, the software solution was just an experiment and it had low priority, so we never merged that into the project.
However, it demonstrated that using some DBMS to store clauses is doable.
This is why I'm pretty sure doing a similar thing with a different DBMS is definitely doable.

**Reply by daniellefisla** [reply]
Thank you for your reply. Sounds like this is doable and what you have shared with me is exactly what I needed to get started. I will take a look at those interfaces and will let you should I implement something usable. Cheers, D.

**Reply by gciatto** [reply]
Nice to read that! 

Feel free to open new discussions in case you need insights about the design or implementation of any piece of code.
